### DISCUSSION TITLE
When the button is off, the script should determine if power is on/off, when button is on, the power should always be on (until it's turn off)

### DISCUSSION THREAD

**Discussion by MagTun** [arendst/Tasmota#13499] [Q&A] [answered]
The situation in more details: 
My tasmota script turns the power on (or off) when the value retrieved from a json (sent by a sensor) are too high (or low enough). But I would like a way to overwrite the script with the button: 
- when the button is on, the power should always be on, no matter what the value in the json are. The power should stay on, until the button is (manually) turn off. 
- if the button is off, then it's the value in the json that controles the power.
  
Currently my button controles the relay so if I switch it on, the power will turn on... until the script finds a low value (and not until I decide to manually press the button). 

The 2 problems to solve:
- How should I set the button in the configuration module?
- How can I check the button state in my script?

Thanks a lot for your help!

**Comment by barbudor** [comment]
Are you using Scripting language or Rules ?

In Tasmota, a "Button" is a push-button that you press and must release
If you want to use a rocker-switch that alternate between an off and a on positin you must use a "Switch"
If you than use command `SetOption114 1` you will decouple the switch position from the relay and can capture the envents into Rules.
I don't know how to do that if you are using Scripting-Language

**Reply by MagTun** [reply]
I am using a script, this [one](https://github.com/MagTun/Turn-automatically-PM-filter-with-smartplug-esp8266-lufdaten-sensor-/blob/main/script_for_webUI.script)

Do you mean that in the configuration, I should assign the button to a switch, instead of the relay? Would this be enough to decouple the button from the relay? 
![image](https://user-images.githubusercontent.com/15843700/139378794-41033862-5e9b-41e4-b7ff-0b2c918bf9f5.png)

**Reply by rj10rd** [reply]
I have same issue with rule
How solve it.

**Comment by gemu2015** [comment, marked as answer]
you have 2 options:
1. declare GPIO buttons via Tasmota. then you should declare an extra "override" virtual relays with which you force the script to on.   i use this approach with my wallbox (one real relays ("load") and 2 virtual relays) it turns on the wallbox if enough solar power is available. 

<img width="391" alt="Bildschirmfoto 2021-10-29 um 08 04 33" src="https://user-images.githubusercontent.com/11647075/139384035-b54b72c7-537c-44a6-88a0-1fe2ad17aafb.png">

2. declare GPIO buttons with script and define the buttons in script also.  this is more effort but also even more flexible.

**Reply by MagTun** [reply]
Thanks for your reply and sorry for the delay. it's good to know that those two options are possible but in myc ase they aren't a good solution because I need to take the control over the plug via a physical mean (most likely the button) and not via the UI interface. 
I though it would have been possible to redirect the button to something else than the relay that turn the plug on/off. That would be a great feature for the next tasmota updates

**Reply by gemu2015** [reply]
that is absolutely possible with a script. you can define the buttons and the plugs relays in script not in Tasmota and then 
handle all actions you want.

**Reply by MagTun** [reply]
Thanks for your reply and for giving me the right terms to search for. 

So I read in the [scripting](https://tasmota.github.io/docs/Scripting-Language/) documentation that I need to define `#define USE_BUTTON_EVENT` which I did and I should use the `>b` section. But the documentation is really minimal after that. 

As I have a a `>S` and a `>E` section, I have added the `>d` section right above it. 

Then in this  `>d`  I tried first to get the current state of the button but even this didn't work. Nothing is printed as if the `>d` section wasn't called. I also wonder which index I should use for  `bt`. 

Here are the relevant part of my script

```
>D
... some var...

>d
buttonstate=bt[0]
print button %buttonstate%

>S
; just to know when S section is called
print SectionS
res=http(ipa "/data.json")

>E
pm10=gwr("value" 4)
pm25=gwr("value" 6)
```

**Reply by gemu2015** [reply]
this example shows how to control hardware with script alone
do not define button or relay in tasmota UI
assumes pin 12 as hardware button
and pin 13 as relay
needs #define USE_SCRIPT_WEB_DISPLAY
just to give an idea of the principle 


\>D
but=0
rel=0
\>B
; set pin 12 to input (button)
spinm(12 I)
; set pin 13 to output (relay)
spinm(13 O)

\>S
; button pressed
if but==0
then
; invert relay state
rel^=1
endif

\>F
; read button state
but=pin(12)

; set relay according to rel state
spin(13 rel)

\>W
; show button on webui
bu(rel "relay on" "relay off")

**Reply by MagTun** [reply]
Thanks for this @gemu2015. 
Based on my current configuration, I think that I can't use the pin 12 and 13 for button / relay. I would say that button is pin 3 and the relay is pin 14 (or is it 5?), but I am not sure. I am afraid that if I set the wrong pin number, I could break the plug, so I just would like to confirm with you the pins I should use, before trying your code. 


![image](https://user-images.githubusercontent.com/15843700/140187557-1389df4c-6bc7-4dcc-83c0-ae263da2aa83.png)

**Reply by gemu2015** [reply]
this was only meant as a general example to show how i works in principle!
it is much more needed to implement your functionality.
with your hardware you should select 3 and 14

**Reply by MagTun** [reply]
Thanks for you help. I have tried your code with the value 3 for the button and 14 for the relay but I am getting this error:
![image](https://user-images.githubusercontent.com/15843700/140413231-9bcf87bd-5a58-4272-9a4a-9b642c64768f.png)

I guess this comes from these lines:

```
>F
; read button state
but=pin(3)
```
Even though I get this error, the button keeps switching on and off every other second or so. 

Also, I have disassociated pin 3 and 14 in config (but kept the other ones):
![image](https://user-images.githubusercontent.com/15843700/140464508-ed41b834-0c74-48c4-8819-0441e31f5785.png)

**Reply by gemu2015** [reply]
sorry typo  must be:
but=pin[3]

but try your modified script here.  i think it is a bit easier.
assumes 2 relays and 2 buttons defined in tasmota webui.
button1 on gpio0 and relay1 gpio1 (virtual relay)
button2 on gpio3 and relay2 on gpio14
needs #define USE_SCRIPT_WEB_DISPLAY
for showing pm values in WEBUI

[script_for_webUI.script.txt](https://github.com/arendst/Tasmota/files/7481425/script_for_webUI.script.txt)

**Reply by MagTun** [reply]
Thanks a lot for digging into my script and proposing a solution. You are amazing! I tried this last script, and the button works as expected but when the button is off and there have been 2 high values, the plug doesn't start. Here is a log

![image](https://user-images.githubusercontent.com/15843700/140575561-5df7c452-df4f-4f8a-8dce-1daaa63f19a4.png)




---
I have tried to mix my script with the one you proposed 2 days ago. This new script is [here](https://pastebin.com/CN5fZVPz), but this one also doesn't work. The button doesn't turn on/off the plug and the button on/off in the UI doesn't seems to work either. Also the led stays on permanently. 

  
For my new script I have disassociated the GPIO 3 and 14 in config but not for yours as I don't see the GPIO declared in it:
![image](https://user-images.githubusercontent.com/15843700/140576155-d51a7245-e44a-4f83-9f8a-5ef773595b06.png)

**Reply by gemu2015** [reply]
if you set GPIOs with script tasmota does not recognize that. you have to manage all button handling and WEBUI in script. this is only recommended for experienced users.
i modified your script again and now decoupled the filter checking completely from the relay.
i think it shall now work as expected.
again define GPIOs in Tasmota like explained above


[script_for_webUI.script.txt](https://github.com/arendst/Tasmota/files/7489993/script_for_webUI.script.txt)

**Reply by MagTun** [reply]
Thanks for you help. Sorry for the GPIO I misunderstood what you wrote. 

I tried you last script. The buttons (the physical one and the UI one) work as expected. The plug turns on when the values are high (and the force is off), but to get that I had to change `power2` into `power0` (or `power1`) as with power2 the plug never turns on when the values are high (cf the last printscreen). 

But now, th plug doesn't stops on its own when the values have become low. I need to manually click the physical button or the UI one to make it stop. 

![image](https://user-images.githubusercontent.com/15843700/140619515-ca8d8ae9-de06-4ec9-8630-a6f97b643cc5.png)


There is a command error when using power2. 


![image](https://user-images.githubusercontent.com/15843700/140619511-b1f753b4-bb75-4a16-b336-934ba31fe4f0.png)

**Reply by gemu2015** [reply]
since you got a command error with power2 you did not configure tasmota gpios as needed
with the last script i posted you need this configuration.
you need a virtual button and relays and a real one.
i checked this script with my device and there is no command error.


<img width="499" alt="Bildschirmfoto 2021-11-07 um 07 55 23" src="https://user-images.githubusercontent.com/11647075/140635606-e6688344-4997-4b81-b1a5-28b831c6d939.png">

**Reply by MagTun** [reply]
Thanks a lot for that, I was really confused with the configuration. 

With this updated configuration, I don't have any power2 error but there are still some bugs.

When the plug is OFF (because the values are low), and I press on the physical button, the plug turns ON as expected. But if I press once again on the physical button, the plug doesn't stop, even though in console power1 and power2 are shown as OFF and the buttons in the UI are shown as OFF . This also happens if I use in the same way the button "force" inside the UI. In both case, the plug waits until it get the next low value to turn itself OFF . 

What  I expected was that the plug would switch OFF when I turn the "force mode" OFF , and then on the next measure, the plug will "decide", based on the values, if it needs to turn on or not. The problem with this setup is that there is not way to know if the force mode has really been turn OFF (without looking at the UI) [I could easily press twice instead of once on the button]. 
But this setup could be kept has is it if there were a way to know via the led when the plug is in 'force mode". Would it be possible to make it flash when the force mode is ON and to make is stay continuously when the plug is ON  because of the high values?

This is after having turn off the "force" mode (even though it shows power OFF, the plug is still ON): 
![image](https://user-images.githubusercontent.com/15843700/140655080-f4e951e2-7fe4-4d6c-b0f3-3cea174bc9e5.png)

![image](https://user-images.githubusercontent.com/15843700/140655277-af6c7c66-039d-4bbb-8af5-31d2460dfdc7.png)

This is after the next values have been measured, and as they are low, the plug turns OFF (the plug was ON from 18:29:05 to 18:30:49 the plug was still ON)  : 
![image](https://user-images.githubusercontent.com/15843700/140655350-e950f930-db34-4f42-8b6d-52ae4c59e419.png)

**Reply by gemu2015** [reply]
yes see what the problem is.
try this modification.  now we check the change of force button and when disabling switch relays is turned off when pm values are OK

[script_for_webUI.script.txt](https://github.com/arendst/Tasmota/files/7492813/script_for_webUI.script.txt)

**Reply by MagTun** [reply]
You are really amazing! 
I tried your last script and did just a little change: when I turn OFF the force mode the plug will now switch OFF (giving me the signal that the force mode is OFF). It will then start by itself if the values are high on the next measurement. 
For this I have simply commented these lines 60-62: 

```
; force off
;if spwr==0
;then
```

I can't thank you enough for all the time you spent for this and for your patience with my misunderstandings. I will share this updated script with the users of Luftdaten , an open source project for measuring the air pollution (PM).

**Reply by MagTun** [reply]
Here is the final [script](https://github.com/MagTun/Smartplug-esp8266-turn-automatically-DIY-PM-filter-with-lufdaten-sensor) with the force mode in the project to get a cheap and efficient DIY PM Filter automatically turn On/OFF based on the open-source [Lufdaten PM sensor](https://sensor.community/en/sensors/airrohr/). 

I credited you for your help! And again, thanks so much @gemu2015
